src/Microsoft.ServiceFabric.Client.Http/ImageStoreClient_.cs

```diff
@@ -1,27 +1,51 @@
-﻿// ------------------------------------------------------------
+// ------------------------------------------------------------
 // Copyright (c) Microsoft Corporation.  All rights reserved.
 // Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
 // ------------------------------------------------------------
 
-using System.Net.Http.Headers;
-
-
 namespace Microsoft.ServiceFabric.Client.Http
 {
     using System;
+    using System.Collections.Concurrent;
     using System.Collections.Generic;
     using System.IO;
+    using System.IO.Compression;
+    using System.Linq;
     using System.Net.Http;
-    using System.Text;
+    using System.Net.Http.Headers;
     using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.ServiceFabric.Client;
-    using Microsoft.ServiceFabric.Client.Http.Serialization;
+    using Microsoft.ServiceFabric.Client.Exceptions;
     using Microsoft.ServiceFabric.Common;
-    using Newtonsoft.Json;
+    using Microsoft.ServiceFabric.Common.Utilities;
 
-    internal partial class ImageStoreClient
+    /// <summary>
+    /// Class containing methods for performing ImageStoreClient operataions.
+    /// </summary>
+    internal partial class ImageStoreClient : IImageStoreClient
     {
+        private string imageStorePath;
+        private const int UploadLimitSizeInBytes = 25 * 1024 * 1024;
+        private const int MaxConcurrentUpload = 10;
+        private const int MaxUploadTry = 2;
+        private const string ZipExtension = "zip";
+        private const long ServerTimeOutInSecondsForUpload = 60;
+
+        private async Task<string> GetImageStoreConnectionString()
+        {
+            if (imageStorePath == null)
+            {
+                imageStorePath = await httpClient.Cluster.GetImageStoreConnectionString();
+                if (IsLocalStore(imageStorePath))
+                    imageStorePath = new Uri(imageStorePath).LocalPath;
+            }
+
+            return imageStorePath;
+        }
+
+        private bool IsLocalStore(string imageStorePath) => imageStorePath != "fabric:ImageStore" && !imageStorePath.StartsWith("xstore");
```

Comment:
These 2 can be merged into on property called IsLocalStore which conditionally does httpClient.Cluster.GetImageStoreConnectionString() and then just compare

Reply:
That would force it to be a Task<bool> property because GetImageStoreConnectionString() is an async method. Property getters can't be declared async so that would make for a bit of a clunky implementation..